vscode-lean4/src/infoview.ts

```diff
@@ -3,16 +3,44 @@ import {
     commands, Disposable, DocumentSelector,
     ExtensionContext, languages, Range,
     Selection, TextEditor, TextEditorRevealType,
-    Uri, ViewColumn, WebviewPanel, window, workspace, env, Position, Diagnostic,
+    Uri, ViewColumn, WebviewPanel, window, workspace, env, Position,
 } from 'vscode';
-import { EditorApi, InfoviewApi, LeanFileProgressParams, TextInsertKind } from '@lean4/infoview';
+import { EditorApi, InfoviewApi, LeanFileProgressParams, TextInsertKind, RpcConnectParams, RpcConnected, RpcKeepAliveParams } from '@lean4/infoview';
 import { LeanClient } from './leanclient';
 import { getInfoViewAllErrorsOnLine, getInfoViewAutoOpen, getInfoViewAutoOpenShowGoal,
     getInfoViewFilterIndex, getInfoViewStyle, getInfoViewTacticStateFilters } from './config';
 import { Rpc } from './rpc';
-import { PublishDiagnosticsParams } from 'vscode-languageclient';
 import * as ls from 'vscode-languageserver-protocol'
 
+const keepAlivePeriodMs = 10000
+
+async function rpcConnect(client: LeanClient, uri: ls.DocumentUri): Promise<string> {
+    const connParams: RpcConnectParams = { uri };
+    const result: RpcConnected = await client.sendRequest('$/lean/rpc/connect', connParams);
+    return result.sessionId;
+}
+
+class RpcSession implements Disposable {
+    keepAliveInterval?: NodeJS.Timeout;
+
+    constructor(client: LeanClient, public sessionId: string, public uri: ls.DocumentUri) {
+        this.keepAliveInterval = setInterval(() => {
+            const params: RpcKeepAliveParams = { uri, sessionId }
+            try {
+                client.sendNotification('$/lean/rpc/keepAlive', params)
```

Comment:
I wonder if this client.sendNotification could overlap with another "sendClientNotification" in another thread? Does it matter?  I wonder if "sendClientNotification" should clearInterval and restart the countdown so that a "keepAlive" doesn't slow down whatever work "sendClientNotification" is trying to do ?

Reply:
Notifications are generally asynchronous, so it could overlap, but I don't think there's any issue with that. The LSP protocol is designed to be asynchronous.

Reply:
Previously using the browser window.setTimeout it would not overlap with keystroke handling, that's why I ask.  This is a new multithreaded overlap potential.

Reply:
I am confused, which keystroke handlers are you referring to? Btw, afaik async JS methods do not run in parallel on multiple threads, although they may interleave.

Reply:
Sorry, the setTimeout before was a web browser call in infoview, which I believe runs on the UI thread, the same thread that would be processing render operations that result in calls to sendClientNotification.  (Those sendClientNotification calls are usually the result of a a mouse hover or move cursor which also happens on keyboard input like "." and therefore I believe previous sendClientNotification versus keepAlive could not be interleaved).  But if you are saying LSP is already fully multithread safe then this is not an issue.

Reply:
I see, I think the same kind of negligibility argument applies here. Looking at the traffic of a typical LSP session (say by enabling server logging) reveals tens of requests per second made by the editor when making changes. One keepalive is really just lost in the woods here. Note also that only RPC requests keep the RPC session alive. `didChange` sent on keystrokes does not. This is intentional since we don't need a session when the infoview is closed.

Reply:
The keepAlive notification won't make a difference here.  With either implementation, the notification can come between a completion request and it's response.  It shouldn't cause any problems,  (You can also have multiple requests in flights btw.) 

All LSP events are routed through the extension which runs in the extension host.  The only thing that changes is where the notification is "started": i.e., infoview->extension->lean vs. extension->lean.  Since you're asking about completions: these go through the extension as well, i.e. editor->extension->lean.